src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java

```diff
@@ -0,0 +1,447 @@
+        TokenTypes.LITERAL_FOR,
+        TokenTypes.LITERAL_WHILE,
+    };
+
+    /**
+     * An array of unacceptable children of ast of type {@link TokenTypes#DOT}.
+     */
+    private static final int[] UNACCEPTABLE_CHILD_OF_DOT = {
+        TokenTypes.DOT,
+        TokenTypes.METHOD_CALL,
+        TokenTypes.LITERAL_NEW,
+        TokenTypes.LITERAL_SUPER,
+        TokenTypes.LITERAL_CLASS,
+        TokenTypes.LITERAL_THIS,
+    };
+
+    /**
+     * An array of unacceptable parent of ast of type {@link TokenTypes#IDENT}.
+     */
+    private static final int[] UNACCEPTABLE_PARENT_OF_IDENT = {
+        TokenTypes.VARIABLE_DEF,
+        TokenTypes.DOT,
+        TokenTypes.LITERAL_NEW,
+        TokenTypes.PATTERN_VARIABLE_DEF,
+        TokenTypes.METHOD_CALL,
+        TokenTypes.TYPE,
+    };
+
+    /**
+     * Keeps VariableDesc objects inside stack of declared variables.
+     */
+    private Deque<VariableDesc> variables;
+
+    @Override
+    public int[] getDefaultTokens() {
+        return new int[] {
+            TokenTypes.DOT,
+            TokenTypes.VARIABLE_DEF,
+            TokenTypes.IDENT,
+            TokenTypes.SLIST,
+            TokenTypes.LITERAL_IF,
+            TokenTypes.LITERAL_FOR,
+            TokenTypes.LITERAL_WHILE,
+        };
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getDefaultTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return getDefaultTokens();
+    }
+
+    @Override
+    public void beginTree(DetailAST root) {
+        variables = new ArrayDeque<>();
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        final int type = ast.getType();
+        if (type == TokenTypes.DOT) {
+            visitDotToken(ast);
+        }
+        else if (type == TokenTypes.VARIABLE_DEF) {
+            visitVariableDefToken(ast);
+        }
+        else if (type == TokenTypes.IDENT) {
+            visitIdentToken(ast);
+        }
+    }
+
+    @Override
+    public void leaveToken(DetailAST ast) {
+        if (TokenUtil.isOfType(ast, SCOPES)) {
+            while (!variables.isEmpty() && variables.peek().getScope() == ast) {
+                final VariableDesc variableDesc = variables.pop();
+                if (!variableDesc.isUsed()) {
+                    final DetailAST typeAst = variableDesc.getTypeAst();
+                    log(typeAst, MSG_UNUSED_LOCAL_VARIABLE, variableDesc.name);
+                }
+            }
+        }
+    }
+
+    /**
+     * Visit ast of type {@link TokenTypes#DOT}.
+     *
+     * @param dotAst dotAst
+     */
+    private void visitDotToken(DetailAST dotAst) {
+        final boolean checkIdentToken = !TokenUtil.findFirstTokenByPredicate(dotAst,
+                        ast -> {
+                            return TokenUtil.isOfType(ast,
+                                    UNACCEPTABLE_CHILD_OF_DOT);
+                        })
+                .isPresent();
+        if (checkIdentToken) {
```

Comment:
it is better to use Optional#ifPresent

Reply:
ifPresent won't quite work in an intended way, the condition with `ifPresent(..)` would be something like
```
TokenUtil.findFirstTokenByPredicate(dotAst,
                        ast -> {
                            return !TokenUtil.isOfType(ast,
                                    UNACCEPTABLE_CHILD_OF_DOT);
                        })
                .ifPresent(ast -> checkIdentifierAst(dotAst.findFirstToken(TokenTypes.IDENT)));
```
I don't like this approach as first of all it doesn't work in this case and secondly, the lambda parameter is not used.
Why this doesn't work is due to the fact that `TokenUtil.findFirstTokenByPredicate(..)` stops when it finds the suitable token i.e. in this case `Every possible token - UNACCEPTABLE_CHILD_OF_DOT` and then it will call the method `checkIdentifierAst(..)`. 
Earlier it stopped when it found `UNACCEPTABLE_CHILD_OF_DOT`. Why this makes a difference can be explained by a simple example-
```
Class<Integer> b = int.class;
```
parses as (the portion we need)-
```
--EXPR -> EXPR
   `--DOT -> .
       |--LITERAL_INT -> int
       `--LITERAL_CLASS -> class
```
Now the method `TokenUtil.findFirstTokenByPredicate(..)` checks by traversing from sibling to sibling and will stop at `LITERAL_INT` as that comes under `Every possible token - UNACCEPTABLE_CHILD_OF_DOT` and will invoke the method `checkIdentifierAst(dotAst.findFirstToken(TokenTypes.IDENT))` which will give an NPE as `DOT` dosen't have a child of type `IDENT`.  

Originally what it's doing is traversing until it finds `UNACCEPTABLE_CHILD_OF_DOT` and then stop. I negated the whole boolean condition and checked the negated condition before calling `checkIdentifierAst(..)`.

So `ifPresent(..)` won't work but `ifPresentOrElse(..)` will work but we can't use it as it's Java 9+